ouroboros-consensus/src/unstable-consensus-testlib/Test/Util/HeaderValidation.hs

```diff
@@ -0,0 +1,49 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE TypeSynonymInstances #-}
+
+module Test.Util.HeaderValidation (
+    -- * Bogus time
+    AddBogusTime
+  , addBogusTime
+  , addBogusTimeToFragment
+  , dropTimeFromFragment
+  ) where
+
+import           Cardano.Slotting.Time (RelativeTime (..))
+import           Data.Typeable (Typeable)
+import           Ouroboros.Consensus.Block (Header)
+import           Ouroboros.Consensus.HeaderValidation (HeaderWithTime (..))
+import           Ouroboros.Network.AnchoredFragment (AnchoredFragment)
+import qualified Ouroboros.Network.AnchoredFragment as AF
+
+{-------------------------------------------------------------------------------
+  Bogus time
+-------------------------------------------------------------------------------}
+
+-- REVIEW: I'm not sure about the name of this type class, and how safe it is to use it in testing code. What if suddenly a test decides to rely on the header's time?
+class AddBogusTime blk where
+  addBogusTime ::
+       Header blk
+    -> HeaderWithTime blk
+
+-- REVIEW: I even wonder if this instance should be defined here and
+-- at all: there might be use cases in which the 'TestBlock's are used
+-- in tests that depend on the header having a meaningful time
+-- associated with them.
```

Comment:
Hmm, one idea: All tests where you are currently adding bogus times are using non-HFC blocks. For these, time translation is very easy as there just is a constant slot length, so this wouldn't even be a "bogus" time!
```haskell
attachSlotTime ::
     NoHardForks blk
  => TopLevelConfig blk
  -> Header blk
  -> HeaderWithTime blk
attachSlotTime cfg hdr = HeaderWithTime {
      hwtHeader           = hdr
    , hwtSlotRelativeTime =
        runIdentity $ epochInfoSlotToRelativeTime ei (blockSlot hdr)
    }
  where
    ei = noHardForksEpochInfo cfg
```
The important thing here is that we can get a *single* `EraParams` from the `TopLevelConfig`. `NoHardForks` provides that via `getEraParams`, but it has `SingleEraBlock` as a superclass constraint, which our `TestBlock`s don't implement (they could, but it requires implementing lots of "boring" stuff so nobody has done that yet). To sidestep that, we could move `getEraParams` to a separate class that doesn't demand `SingleEraBlock` instead.